### PR TITLE
Expose replication lag metrics

### PR DESCRIPTION
**Pull request by royjacobson**
Part of #1259.

### CODE REVIEW

File: src/server/dflycmd.h [hunk 1]
```diff
@@ -214,4 +206,6 @@ class DflyCmd {
   Mutex mu_;  // Guard global operations. See header top for locking levels.
 };
 
+std::string_view SyncStateName(DflyCmd::SyncState sync_state);
```
Comment (chakaz) [inline src/server/dflycmd.h:209]: Wouldn't it be more elegant to place this function within `DflyCmd`, next to the definition for `SyncState` itself?

File: src/server/server_family.cc [hunk 2]
```diff
@@ -869,6 +869,19 @@ void PrintPrometheusMetrics(const Metrics& m, StringResponse* resp) {
                       &db_key_expire_metrics);
   }
 
+  if (!m.replication_metrics.empty()) {
```
Comment (chakaz) [inline src/server/server_family.cc:872]: Can we test this functionality in any way? Either unit tests or pytests?

File: src/server/dflycmd.cc [hunk 3]
```diff
@@ -63,24 +65,18 @@ struct TransactionGuard {
...
+      return "preparation";
+    case DflyCmd::SyncState::FULL_SYNC:
+      return "full_sync";
+    case DflyCmd::SyncState::STABLE_SYNC:
+      return "stable_sync";
+    case DflyCmd::SyncState::CANCELLED:
+      return "cancelled";
     default:
-      break;
+      LOG(FATAL) << "bad state " << int(sync_state);
```
[the remark this answers is not in the document]
Comment (royjacobson) [reply src/server/dflycmd.cc:79]: We already turn on `-Wswitch`, actually :) So yeah, let's remove the `default:`.

File: src/server/server_family.cc (hunk 2 above)
Comment (royjacobson) [reply src/server/server_family.cc:872]: We should really do that.. We have no coverage for the metrics at all right now.

**Review by chakaz — DISMISSED** [review]

File: src/server/dflycmd.h [hunk 4]
```diff
@@ -91,8 +92,9 @@ struct FlowInfo {
 //
 class DflyCmd {
  public:
-  // See header comments for state descriptions.
+  // See class comments for state descriptions.
   enum class SyncState { PREPARATION, FULL_SYNC, STABLE_SYNC, CANCELLED };
+  static std::string_view SyncStateName(DflyCmd::SyncState sync_state);
```
Comment (adiholden) [inline src/server/dflycmd.h:97]: I dont see we are calling this function from outside dflycmd.cc so we can move it from the header and move the impl to the unnamed namespace

File: src/server/replica.h [hunk 5]
```diff
@@ -28,6 +28,13 @@ class ConnectionContext;
 class JournalExecutor;
 struct JournalReader;
 
+struct ReplicaRoleInfo {
```
Comment (adiholden) [inline src/server/replica.h:31]: The struct is used to return the data from GetReplicasRoleInfo. I dont think this is the right place for this struct.
This is the data master holds on its replica. 
Can we move it back to dflycmd.h ?

File: src/server/dflycmd.cc [hunk 6]
```diff
@@ -63,25 +65,18 @@ struct TransactionGuard {
...
+    case DflyCmd::SyncState::PREPARATION:
+      return "preparation";
+    case DflyCmd::SyncState::FULL_SYNC:
+      return "full_sync";
+    case DflyCmd::SyncState::STABLE_SYNC:
+      return "stable_sync";
+    case DflyCmd::SyncState::CANCELLED:
+      return "cancelled";
+  }
+  LOG(FATAL) << "Unspported state " << int(sync_state);
```
Comment (adiholden) [inline src/server/dflycmd.cc:79]: Better use log error here.
f.e if you add a new state and start a replica with the new version while master is in the old version were this state is not supported you will get a crash

File: tests/dragonfly/replication_test.py [hunk 7]
```diff
@@ -911,7 +911,22 @@ def parse_lag(replication_info: bytes):
...
+    is satisfied at least once.
+    We check both `INFO REPLICATION` redis protocol and the `/metrics`
+    prometheus endpoint.
+    """
+    for _ in range(10):
+        lag = (await inst.metrics())["dragonfly_connected_replica_lag_records"].samples[0].value
+        if condition(lag):
+            break
+        print("current prometheus lag =", lag)
+        time.sleep(0.05)
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:927]: use await asyncio.sleep

File: src/server/dflycmd.cc (hunk 6 above)
Comment (royjacobson) [reply src/server/dflycmd.cc:79]: It's controlled by the state machine code in the same binary, it's not input from the other client.

File: src/server/replica.h (hunk 5 above)
Comment (royjacobson) [reply src/server/replica.h]: I need it to be defined in server_family.h when I include it in metrics though

Comment (royjacobson) [reply src/server/replica.h]: Tried to move it to server_family.h (slightly better, I think)

File: src/server/dflycmd.cc (hunk 6 above)
Comment (adiholden) [reply src/server/dflycmd.cc]: you are right. I would still crash in debug mode and in release just return unsupported

Comment (adiholden) [reply src/server/dflycmd.cc]: ping here

**Review by adiholden — APPROVED** [review]

**Review by adiholden — APPROVED** [review]

**Pull request merged**